# Import libraries

In [1]:
import pandas as pd 
import numpy as np
import os 
import requests
import json
from datetime import datetime

# Concatenate the files

In [2]:
# Create a list with all files in the folder using a list comprehension
folderpath = r"Data"
filepaths  = [os.path.join(folderpath, name) for name in os.listdir(folderpath)]

In [3]:
filepaths

['Data\\202201-citibike-tripdata_1.csv',
 'Data\\202201-citibike-tripdata_2.csv',
 'Data\\202202-citibike-tripdata_1.csv',
 'Data\\202202-citibike-tripdata_2.csv',
 'Data\\202203-citibike-tripdata_1.csv',
 'Data\\202203-citibike-tripdata_2.csv',
 'Data\\202204-citibike-tripdata_1.csv',
 'Data\\202204-citibike-tripdata_2.csv',
 'Data\\202204-citibike-tripdata_3.csv',
 'Data\\202205-citibike-tripdata_1.csv',
 'Data\\202205-citibike-tripdata_2.csv',
 'Data\\202205-citibike-tripdata_3.csv',
 'Data\\202206-citibike-tripdata_1.csv',
 'Data\\202206-citibike-tripdata_2.csv',
 'Data\\202206-citibike-tripdata_3.csv',
 'Data\\202206-citibike-tripdata_4.csv',
 'Data\\202207-citibike-tripdata_1.csv',
 'Data\\202207-citibike-tripdata_2.csv',
 'Data\\202207-citibike-tripdata_3.csv',
 'Data\\202207-citibike-tripdata_4.csv',
 'Data\\202208-citibike-tripdata_1.csv',
 'Data\\202208-citibike-tripdata_2.csv',
 'Data\\202208-citibike-tripdata_3.csv',
 'Data\\202208-citibike-tripdata_4.csv',
 'Data\\202209-c

In [4]:
# Read and concatenate all files, setting low_memory=False
df = pd.concat((pd.read_csv(f, low_memory=False) for f in filepaths), ignore_index=True)

In [5]:
# Take a 1% sample of the data to reduce size
df_sample = df.sample(frac=0.01, random_state=42)

In [6]:
df.shape

(29838806, 13)

In [7]:
df_sample.shape

(298388, 13)

In [8]:
df_sample.head()

ride_id  rideable_type               started_at  \
15115901  25098D3FFB6B3BBD  electric_bike  2022-07-04 17:00:41.408   
753105    FCED437CC64165BA   classic_bike  2022-01-14 09:18:35.838   
23821092  22DE76A19BAD1251  electric_bike  2022-10-09 17:35:17.370   
17525642  1A009107ED10A6FB  electric_bike  2022-08-05 19:10:15.391   
6397983   2B63FBA517507650   classic_bike  2022-05-10 15:28:01.330   

                         ended_at           start_station_name  \
15115901  2022-07-04 17:19:00.525  Central Park West & W 85 St   
753105    2022-01-14 09:27:58.570             W 4 St & 7 Ave S   
23821092  2022-10-09 18:08:06.487   Menahan St & Onderdonk Ave   
17525642  2022-08-05 19:38:10.967          Hope St & Union Ave   
6397983   2022-05-10 15:29:29.183              6 Ave & W 33 St   

         start_station_id                   end_station_name end_station_id  \
15115901          7354.01  Grand Army Plaza & Central Park S        6839.10   
753105            5880.02            W 16 St & The High Line        6233.05   
23821092          5034.03              N 12 St & Bedford Ave        5450.04   
17525642          5187.03                     38 St & 30 Ave        6850.01   
6397983           6364.07                 Broadway & W 29 St        6289.06   

          start_lat  start_lng    end_lat    end_lng member_casual  
15115901  40.784760 -73.969862  40.764397 -73.973715        casual  
753105    40.734011 -74.002939  40.743349 -74.006818        member  
23821092  40.705580 -73.909559  40.720798 -73.954847        member  
17525642  40.711662 -73.951458  40.764175 -73.915840        member  
6397983   40.749013 -73.988484  40.746201 -73.988557        member

In [9]:
df_sample.tail()

ride_id rideable_type               started_at  \
14132014  CC97820FF0AD109B  classic_bike  2022-07-21 12:14:41.547   
19133558  1DF1B7E1CB6EC263  classic_bike  2022-08-21 15:48:27.920   
1795753   F80578D8DFB6AB73  classic_bike  2022-02-23 17:28:27.165   
22023644  B13119FBEA968D3E  classic_bike  2022-09-17 12:52:11.576   
17998959  1A0C59AAA7700645  classic_bike  2022-08-20 20:17:46.150   

                         ended_at       start_station_name start_station_id  \
14132014  2022-07-21 12:19:33.749       E 5 St & Cooper Sq          5712.12   
19133558  2022-08-21 15:53:08.397  Bank St & Washington St          5964.01   
1795753   2022-02-23 17:33:38.657          W 21 St & 6 Ave          6140.05   
22023644  2022-09-17 12:54:28.582          E 31 St & 3 Ave          6239.08   
17998959  2022-08-20 20:33:16.334     Suydam St & Broadway          4689.03   

                       end_station_name end_station_id  start_lat  start_lng  \
14132014              E 2 St & Avenue B        5515.02  40.727690 -73.990993   
19133558               W 4 St & 7 Ave S        5880.02  40.736197 -74.008592   
1795753                 8 Ave & W 27 St        6297.07  40.741740 -73.994156   
22023644                2 Ave & E 29 St        6122.09  40.743943 -73.979661   
17998959  Metropolitan Ave & Meeker Ave        5300.05  40.695440 -73.932230   

            end_lat    end_lng member_casual  
14132014  40.722174 -73.983688        member  
19133558  40.734011 -74.002939        member  
1795753   40.747968 -73.996637        member  
22023644  40.741724 -73.978093        member  
17998959  40.714133 -73.952344        member

### Code Explanation

1. **Filepath List Creation**:
   - The code uses a list comprehension to create a list of file paths for all files in the `Data` folder.
   - It combines the folder path (`folderpath`) with each file name in the directory (`os.listdir(folderpath)`).

2. **File Reading and Concatenation**:
   - Reads each CSV file from the list of file paths using `pd.read_csv`.
   - Combines all the data into a single DataFrame using `pd.concat`.
   - The `low_memory=False` parameter ensures better handling of mixed data types in columns.

3. **Sampling**:
   - Reduces the size of the combined dataset by randomly selecting 1% of the rows using `df.sample`.
   - The parameter `frac=0.01` specifies the fraction of data to sample.
   - The `random_state=42` ensures the sampling is reproducible, meaning the same subset will be selected if the code is rerun.

This approach processes multiple large files and reduces the dataset size to stay within limits for further processing or sharing.

# Obtain weather data from New York LaGuardia’s weather station for 2022

In [10]:
Token = 'mMBrfpZXuBErfwBxbXzvZTUfTMjjapWF'

In [11]:
# Get the API 
r = requests.get('https://www.ncdc.noaa.gov/cdo-web/api/v2/data?datasetid=GHCND&datatypeid=TAVG&limit=1000&stationid=GHCND:USW00014732&startdate=2022-01-01&enddate=2022-12-31', headers={'token':Token})

In [12]:
# Load the api response as a json
d = json.loads(r.text)

In [13]:
d

{'metadata': {'resultset': {'offset': 1, 'count': 365, 'limit': 1000}},
 'results': [{'date': '2022-01-01T00:00:00',
   'datatype': 'TAVG',
   'station': 'GHCND:USW00014732',
   'attributes': 'H,,S,',
   'value': 116},
  {'date': '2022-01-02T00:00:00',
   'datatype': 'TAVG',
   'station': 'GHCND:USW00014732',
   'attributes': 'H,,S,',
   'value': 114},
  {'date': '2022-01-03T00:00:00',
   'datatype': 'TAVG',
   'station': 'GHCND:USW00014732',
   'attributes': 'H,,S,',
   'value': 14},
  {'date': '2022-01-04T00:00:00',
   'datatype': 'TAVG',
   'station': 'GHCND:USW00014732',
   'attributes': 'H,,S,',
   'value': -27},
  {'date': '2022-01-05T00:00:00',
   'datatype': 'TAVG',
   'station': 'GHCND:USW00014732',
   'attributes': 'H,,S,',
   'value': 32},
  {'date': '2022-01-06T00:00:00',
   'datatype': 'TAVG',
   'station': 'GHCND:USW00014732',
   'attributes': 'H,,S,',
   'value': 49},
  {'date': '2022-01-07T00:00:00',
   'datatype': 'TAVG',
   'station': 'GHCND:USW00014732',
   'attribut

In [14]:
# Secure all items in the response that correspond to TAVG
avg_temps = [item for item in d['results'] if item['datatype']=='TAVG']

In [15]:
# Get only the date field from all average temperature readings
dates_temp = [item['date'] for item in avg_temps]

In [16]:
# Get the temperature from all average temperature readings
temps = [item['value'] for item in avg_temps]

In [17]:
temps

[116,
 114,
 14,
 -27,
 32,
 49,
 7,
 -25,
 14,
 16,
 -54,
 -19,
 40,
 48,
 -67,
 -80,
 39,
 18,
 32,
 51,
 -60,
 -59,
 -7,
 -2,
 36,
 -23,
 -42,
 1,
 -48,
 -71,
 -34,
 -17,
 23,
 64,
 58,
 -28,
 -48,
 5,
 41,
 28,
 63,
 86,
 118,
 28,
 -43,
 -47,
 16,
 116,
 99,
 1,
 -26,
 41,
 56,
 144,
 11,
 8,
 -11,
 24,
 8,
 26,
 77,
 56,
 -19,
 31,
 96,
 182,
 87,
 32,
 40,
 72,
 51,
 -23,
 53,
 116,
 136,
 82,
 139,
 131,
 143,
 104,
 107,
 72,
 57,
 96,
 104,
 59,
 -15,
 -12,
 26,
 99,
 131,
 72,
 70,
 76,
 97,
 86,
 91,
 128,
 114,
 86,
 88,
 134,
 154,
 163,
 157,
 157,
 83,
 79,
 76,
 94,
 106,
 146,
 133,
 128,
 104,
 115,
 116,
 82,
 104,
 129,
 136,
 121,
 124,
 120,
 172,
 148,
 100,
 104,
 135,
 154,
 161,
 171,
 160,
 179,
 188,
 204,
 197,
 178,
 155,
 154,
 212,
 278,
 210,
 174,
 162,
 164,
 202,
 204,
 207,
 233,
 283,
 178,
 203,
 195,
 219,
 200,
 222,
 218,
 237,
 244,
 218,
 218,
 205,
 242,
 239,
 233,
 203,
 252,
 217,
 173,
 205,
 221,
 191,
 192,
 219,
 258,
 270,
 243,
 22

In [18]:
# Put the results in a dataframe
df_temp = pd.DataFrame()

In [19]:
# Get only date and cast it to date time; convert temperature from tenths of Celsius to normal Celsius
df_temp['date'] = [datetime.strptime(d, "%Y-%m-%dT%H:%M:%S") for d in dates_temp]
df_temp['avgTemp'] = [float(v)/10.0 for v in temps]

In [20]:
df_temp.head()

date  avgTemp
0 2022-01-01     11.6
1 2022-01-02     11.4
2 2022-01-03      1.4
3 2022-01-04     -2.7
4 2022-01-05      3.2

In [21]:
df_temp.tail()

date  avgTemp
360 2022-12-27     -0.7
361 2022-12-28      3.4
362 2022-12-29      6.4
363 2022-12-30      9.3
364 2022-12-31      8.2

In [22]:
df_sample.dtypes

ride_id                object
rideable_type          object
started_at             object
ended_at               object
start_station_name     object
start_station_id       object
end_station_name       object
end_station_id         object
start_lat             float64
start_lng             float64
end_lat               float64
end_lng               float64
member_casual          object
dtype: object

In [23]:
# Convert the 'started_at' column to a datetime format to ensure it is properly recognized as datetime for further analysis or operations.
# This transformation is crucial for performing time-based analysis such as calculating trip duration, filtering by date, or grouping by time periods.
df_sample['started_at'] = pd.to_datetime(df_sample['started_at'])

# Similarly, convert the 'ended_at' column to datetime format for consistency and to enable calculations such as trip duration.
df_sample['ended_at'] = pd.to_datetime(df_sample['ended_at'])

In [24]:
# Explicitly specify the format for 'started_at' column to ensure proper datetime conversion.
# This is useful when you know the exact format of the datetime string to avoid incorrect parsing.
df_sample['started_at'] = pd.to_datetime(df_sample['started_at'], format='%Y-%m-%d %H:%M:%S.%f')

# Explicitly specify the format for 'ended_at' column to ensure consistent datetime conversion.
df_sample['ended_at'] = pd.to_datetime(df_sample['ended_at'], format='%Y-%m-%d %H:%M:%S.%f')

# Alternatively, let pandas infer the format for 'started_at' column.
# This method is flexible but might be slightly slower for large datasets.
df_sample['started_at'] = pd.to_datetime(df_sample['started_at'], infer_datetime_format=True)

# Similarly, let pandas infer the format for 'ended_at' column.
df_sample['ended_at'] = pd.to_datetime(df_sample['ended_at'], infer_datetime_format=True)

C:\Users\anush\AppData\Local\Temp\ipykernel_23088\3412738172.py:10: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  df_sample['started_at'] = pd.to_datetime(df_sample['started_at'], infer_datetime_format=True)
C:\Users\anush\AppData\Local\Temp\ipykernel_23088\3412738172.py:13: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  df_sample['ended_at'] = pd.to_datetime(df_sample['ended_at'], infer_datetime_format=True)


In [25]:
# Extract the date from the 'started_at' column by converting it to datetime and then getting the date part
df_sample['start_date'] = pd.to_datetime(df_sample['started_at'], format='%Y-%m-%d %H:%M:%S.%f').dt.date

# Similarly, extract the date from the 'ended_at' column
df_sample['end_date'] = pd.to_datetime(df_sample['ended_at'], format='%Y-%m-%d %H:%M:%S.%f').dt.date

In [26]:
# Convert the 'date' column to datetime
df_sample['start_date'] = pd.to_datetime(df_sample['start_date'])

# Convert the 'end_date' column to datetime as well
df_sample['end_date'] = pd.to_datetime(df_sample['end_date'])

In [27]:
df_sample.head()

ride_id  rideable_type              started_at  \
15115901  25098D3FFB6B3BBD  electric_bike 2022-07-04 17:00:41.408   
753105    FCED437CC64165BA   classic_bike 2022-01-14 09:18:35.838   
23821092  22DE76A19BAD1251  electric_bike 2022-10-09 17:35:17.370   
17525642  1A009107ED10A6FB  electric_bike 2022-08-05 19:10:15.391   
6397983   2B63FBA517507650   classic_bike 2022-05-10 15:28:01.330   

                        ended_at           start_station_name  \
15115901 2022-07-04 17:19:00.525  Central Park West & W 85 St   
753105   2022-01-14 09:27:58.570             W 4 St & 7 Ave S   
23821092 2022-10-09 18:08:06.487   Menahan St & Onderdonk Ave   
17525642 2022-08-05 19:38:10.967          Hope St & Union Ave   
6397983  2022-05-10 15:29:29.183              6 Ave & W 33 St   

         start_station_id                   end_station_name end_station_id  \
15115901          7354.01  Grand Army Plaza & Central Park S        6839.10   
753105            5880.02            W 16 St & The High Line        6233.05   
23821092          5034.03              N 12 St & Bedford Ave        5450.04   
17525642          5187.03                     38 St & 30 Ave        6850.01   
6397983           6364.07                 Broadway & W 29 St        6289.06   

          start_lat  start_lng    end_lat    end_lng member_casual start_date  \
15115901  40.784760 -73.969862  40.764397 -73.973715        casual 2022-07-04   
753105    40.734011 -74.002939  40.743349 -74.006818        member 2022-01-14   
23821092  40.705580 -73.909559  40.720798 -73.954847        member 2022-10-09   
17525642  40.711662 -73.951458  40.764175 -73.915840        member 2022-08-05   
6397983   40.749013 -73.988484  40.746201 -73.988557        member 2022-05-10   

           end_date  
15115901 2022-07-04  
753105   2022-01-14  
23821092 2022-10-09  
17525642 2022-08-05  
6397983  2022-05-10

In [28]:
df_temp.head()

date  avgTemp
0 2022-01-01     11.6
1 2022-01-02     11.4
2 2022-01-03      1.4
3 2022-01-04     -2.7
4 2022-01-05      3.2

In [29]:
# Export the DataFrame to a CSV file
df_temp.to_csv('temp_data.csv', index=False)

In [30]:
df_temp.dtypes

date       datetime64[ns]
avgTemp           float64
dtype: object

In [31]:
df_sample.dtypes

ride_id                       object
rideable_type                 object
started_at            datetime64[ns]
ended_at              datetime64[ns]
start_station_name            object
start_station_id              object
end_station_name              object
end_station_id                object
start_lat                    float64
start_lng                    float64
end_lat                      float64
end_lng                      float64
member_casual                 object
start_date            datetime64[ns]
end_date              datetime64[ns]
dtype: object

# Merge the weather data with the New York CitiBike data set and export it to a CSV file

In [32]:
# Convert the 'start_date' column in df_sample and 'date' column in df_temp to datetime format
df_sample['start_date'] = pd.to_datetime(df_sample['start_date'])
df_temp['date'] = pd.to_datetime(df_temp['date'])

In [33]:
# Extract the date from the 'start_date' column to use it as a key for merging
df_sample['date'] = df_sample['start_date'].dt.date

In [34]:
%%time

# Convert 'date' column in df_sample to datetime
df_sample['date'] = pd.to_datetime(df_sample['date'], errors='coerce')

CPU times: total: 15.6 ms
Wall time: 45.9 ms


In [35]:
# Check for any rows with NaT and drop them
df_sample = df_sample.dropna(subset=['date'])

In [36]:
# Merge the DataFrames on 'date'
df_merged = df_sample.merge(df_temp, how='left', on='date', indicator=True)

In [37]:
df_merged.head()

ride_id  rideable_type              started_at  \
0  25098D3FFB6B3BBD  electric_bike 2022-07-04 17:00:41.408   
1  FCED437CC64165BA   classic_bike 2022-01-14 09:18:35.838   
2  22DE76A19BAD1251  electric_bike 2022-10-09 17:35:17.370   
3  1A009107ED10A6FB  electric_bike 2022-08-05 19:10:15.391   
4  2B63FBA517507650   classic_bike 2022-05-10 15:28:01.330   

                 ended_at           start_station_name start_station_id  \
0 2022-07-04 17:19:00.525  Central Park West & W 85 St          7354.01   
1 2022-01-14 09:27:58.570             W 4 St & 7 Ave S          5880.02   
2 2022-10-09 18:08:06.487   Menahan St & Onderdonk Ave          5034.03   
3 2022-08-05 19:38:10.967          Hope St & Union Ave          5187.03   
4 2022-05-10 15:29:29.183              6 Ave & W 33 St          6364.07   

                    end_station_name end_station_id  start_lat  start_lng  \
0  Grand Army Plaza & Central Park S        6839.10  40.784760 -73.969862   
1            W 16 St & The High Line        6233.05  40.734011 -74.002939   
2              N 12 St & Bedford Ave        5450.04  40.705580 -73.909559   
3                     38 St & 30 Ave        6850.01  40.711662 -73.951458   
4                 Broadway & W 29 St        6289.06  40.749013 -73.988484   

     end_lat    end_lng member_casual start_date   end_date       date  \
0  40.764397 -73.973715        casual 2022-07-04 2022-07-04 2022-07-04   
1  40.743349 -74.006818        member 2022-01-14 2022-01-14 2022-01-14   
2  40.720798 -73.954847        member 2022-10-09 2022-10-09 2022-10-09   
3  40.764175 -73.915840        member 2022-08-05 2022-08-05 2022-08-05   
4  40.746201 -73.988557        member 2022-05-10 2022-05-10 2022-05-10   

   avgTemp _merge  
0     25.6   both  
1      4.8   both  
2     12.3   both  
3     29.3   both  
4     15.4   both

In [38]:
df_merged['_merge'].value_counts(dropna = False)

_merge
both          298380
left_only          8
right_only         0
Name: count, dtype: int64

### Insights from the Merge

- The vast majority of the rows (298,380) successfully merged with weather data.
- 8 rows in `df_sample` do not have matching weather data for the respective dates. These rows will have `NaN` values in the weather-related columns.
- There are no rows in `df_temp` that lack corresponding CitiBike data, meaning `df_temp` is fully utilized in the join.

In [39]:
df_merged.shape

(298388, 18)

In [40]:
df_merged.to_csv('newyork_citibike_data.csv')